lib/Segments/WooCommerce.php

```diff
@@ -202,260 +236,285 @@ private function needsCollationChange(): bool {
-    ', $subscribersTable, $wpdb->users, $wpdb->usermeta, Source::WOOCOMMERCE_USER));
+    $subscribersTable = $this->entityManager->getClassMetadata(SubscriberEntity::class)->getTableName();
+    $this->connection->executeQuery("
+      UPDATE LOW_PRIORITY {$subscribersTable} mps
+        JOIN {$wpdb->users} wu ON mps.wp_user_id = wu.id
+        JOIN {$wpdb->usermeta} wpum ON wu.id = wpum.user_id AND wpum.meta_key = :capabilities
+      SET is_woocommerce_user = 1, source = :source
+        WHERE wpum.meta_value LIKE '%\"customer\"%'
+    ", ['capabilities' => $wpdb->prefix . 'capabilities', 'source' => Source::WOOCOMMERCE_USER]);
   }
 
-  private function insertSubscribersFromOrders($orderId = null, $status = Subscriber::STATUS_SUBSCRIBED) {
+  private function insertSubscribersFromOrders($orderId = null, $status = Subscriber::STATUS_SUBSCRIBED): array {
     global $wpdb;
-    $subscribersTable = Subscriber::$_table;
+    $validator = new ModelValidator();
     $orderId = !is_null($orderId) ? (int)$orderId : null;
 
-    $insertedUsersEmails = ORM::for_table($wpdb->users)->raw_query(
-      'SELECT DISTINCT wppm.meta_value as email FROM `' . $wpdb->prefix . 'postmeta` wppm
-        JOIN `' . $wpdb->prefix . 'posts` p ON wppm.post_id = p.ID AND p.post_type = "shop_order"
-        WHERE wppm.meta_key = "_billing_email" AND wppm.meta_value != ""
-        ' . ($orderId ? ' AND p.ID = "' . $orderId . '"' : '') . '
-      ')->findArray();
-
-    Subscriber::rawExecute(sprintf('
-      INSERT IGNORE INTO %1$s (is_woocommerce_user, email, status, created_at, last_subscribed_at, source)
-      SELECT 1, wppm.meta_value, "%2$s", CURRENT_TIMESTAMP(), CURRENT_TIMESTAMP(), "%3$s" FROM `' . $wpdb->prefix . 'postmeta` wppm
-        JOIN `' . $wpdb->prefix . 'posts` p ON wppm.post_id = p.ID AND p.post_type = "shop_order"
-        WHERE wppm.meta_key = "_billing_email" AND wppm.meta_value != ""
-        ' . ($orderId ? ' AND p.ID = "' . $orderId . '"' : '') . '
-      ON DUPLICATE KEY UPDATE is_woocommerce_user = 1
-    ', $subscribersTable, $status, Source::WOOCOMMERCE_USER));
+    $subscribersTable = $this->entityManager->getClassMetadata(SubscriberEntity::class)->getTableName();
+    $subQuery = " AND wppm.meta_value NOT IN (
+        SELECT email
+        FROM {$subscribersTable}
+        WHERE is_woocommerce_synced = 1
+        ORDER BY email
+    )";
+    $parameters = ['batchSize' => self::BATCH_SIZE];
+    if ($orderId) {
+      $parameters['orderId'] = $orderId;
+    }
 
-    return $insertedUsersEmails;
-  }
+    $usersEmails = $this->connection->executeQuery('
+      SELECT DISTINCT wppm.meta_value as email FROM `' . $wpdb->prefix . 'postmeta` wppm
+      JOIN `' . $wpdb->prefix . 'posts` p ON wppm.post_id = p.ID AND p.post_type = "shop_order"
+      WHERE wppm.meta_key = "_billing_email" AND wppm.meta_value != ""
+      ' . ($orderId ? ' AND p.ID = :orderId' : $subQuery) . '
+      ORDER BY wppm.meta_value
+      LIMIT :batchSize
```

Comment:
The subquery will return more and more emails as the sync progresses so in the last iteration for a shop with 100K email the query would return 90k emails and MySQL would have to compare the meta_value with all these string in the NOT IN condition.
```sql
AND wppm.meta_value NOT IN (
        SELECT email
        FROM {$subscribersTable}
        WHERE is_woocommerce_synced = 1
        ORDER BY email
    )
  ```
  I have an idea for a different approach for the batching. I think we could use orderId, which is a primary key in wp_post table. We could find the lowestOrderId and  highestOrderId and start fetching using range of orderids.
We could store the last processed orderId in the scheduled task meta.
 
 Here is an example of such a query:
 ```sql
 SELECT DISTINCT wppm.meta_value as email FROM wp_postmeta wppm
      JOIN  wp_posts p ON wppm.post_id = p.ID AND p.post_type = "shop_order"
      WHERE wppm.meta_key = "_billing_email" AND wppm.meta_value != ""
      AND p.ID > 1000 and p.ID < 4000
      ORDER BY wppm.meta_value;
 ```
 With this approcach we could remove  `is_woocommerce_synced` and also the potentialy problematic update in reset at the beginning.

Reply:
Thanks for this idea. These changes make sense to me. I will prepare additional changes.